text/0032-distinct-attribute.md

```diff
@@ -0,0 +1,238 @@
+
+Given this setting :
+```json
+{
+    "distinctAttribute": "product_id"
+}
+```
+
+MeiliSearch will return de-duplicated hits by `product_id`.
+
+```json
+{
+    "hits": [
+        {
+            "colors": "red",
+            "id": 1,
+            "label": "t-shirt",
+            "product_id": 1
+        }
+    ],
+    "offset": 0,
+    "limit": 20,
+    "nbHits": 1,
+    "exhaustiveNbHits": false,
+    "processingTimeMs": 1,
+    "query": "\"t-shirt\""
+}
+```
+
+It returns the top most relevant document by discarding all the others.
+
+For a search with `"q": "black"` as parameter, MeiliSearch returns:
+
+```json
+{
+    "hits": [
+        {
+            "colors": "black",
+            "id": 2,
+            "label": "t-shirt",
+            "product_id": 1
+        }
+    ],
+    "offset": 0,
+    "limit": 20,
+    "nbHits": 1,
+    "exhaustiveNbHits": false,
+    "processingTimeMs": 0,
+    "query": "\"black\""
+}
+```
+
+#### Distinct Attribute in 0.21 (Milli)
+
+Milli shoud be identical as v.0.20 concerning API endpoints and returned results.
+
+### V. Impact on documentation
+N/A
+
+### VI. Impact on SDKs
+N/A
+
+## 2. Technical Aspects
+
+### Abstract
+
+To apply the distinct behavior, the search engine needs to create a database. This database is the same as we create to apply facets and filters on an attribute. However, the users will not pass the attribute into `attributesForFaceting` when setting a distinct attribute. It means the search engine must create this database for the related attribute.
+
+Following this example, it also means the search engine would be technically able to apply filters and facet distribution on the distinct attribute, however, we should prevent this. To avoid confusion, the search engine should prevent the users to execute a filter or get facet distribution on the distinct attribute. Only the distinct capability should be available for this field.
+
+If the user wants to filter on that attribute, he will have to add it in `attributesForFaceting` as well.
+
+Using this new data structure allows interesting future possibilities.
+
+## 3. Future possibilities
+
+> These ideas will be materialized as feature-proposals in the product team backlog. Keep in mind that this specification is related to v0.21.
+
+### Grouping
+#### Distinct count to enable grouping feature
+
+Since MeiliSearch can only de-duplicate documents matching the distinct attribute, it would be interesting to be able to set the number of topmost relevant documents before discarding the others.
+
+`groupLimit` could be applied at search time to fit users needs.
+
+#### Distinct on multiple fields
+
+> It probably require to add a nested structure to return hits for each groups with clarity.
+
+##### Indexing time
+
+Given this distinct setting:
+
+```json
+{
+    "distinctAttributes": ["colors", "size"]
+}
+```
+
+##### Search time
+
+We could add a `groupBy` like parameter to let the user choose a specific grouping behavior:
```

Comment:
groupby would mean that we would have sub-objects on hits ?

```js
hits = {
   red: [ { id: 1, color: "red", label: "nike"}, { id: 2, color: "red", label: "reebok"} ],
   yellow: [ { id: 1, color: "yellow", label: "nike"}, { id: 1, color: "yellow", label: "reebok"} ]
}
```

Reply:
Exactly! See TypeSense group_by feature! https://typesense.org/docs/0.19.0/api/documents.html#group-by

Reply:
Okey! Its a choice ... :')